File: backend/src/lib/storage-paths.ts
import { extname } from "node:path";

export function sanitizeFileName(fileName: string): string {
  const base = fileName.split(/[/\\]/).pop() ?? "upload";
  return base.replace(/[^a-zA-Z0-9._-]/g, "_");
}

export function sourceStoragePath(userId: string, projectId: string, fileName: string): string {
  const safe = sanitizeFileName(fileName);
  return `${userId}/${projectId}/source/${safe}`;
}

export function sourceFileType(mediaType: "audio" | "video"): "source_audio" | "source_video" {
  return mediaType === "audio" ? "source_audio" : "source_video";
}

export function inferMimeType(fileName: string): string | null {
  const ext = extname(fileName).toLowerCase();
  const map: Record<string, string> = {
    ".mp3": "audio/mpeg",
    ".wav": "audio/wav",
    ".m4a": "audio/mp4",
    ".mp4": "video/mp4",
    ".mov": "video/quicktime",
    ".webm": "video/webm",
  };
  return map[ext] ?? null;
}

export const ALLOWED_MIME_TYPES = new Set([
  "audio/mpeg",
  "audio/wav",
  "audio/mp4",
  "audio/x-m4a",
  "video/mp4",
  "video/quicktime",
  "video/webm",
]);

export const MAX_UPLOAD_BYTES = 500 * 1024 * 1024;

File: backend/src/captionovo/storage_paths.py
from pathlib import Path

ALLOWED_MIME_TYPES = {
    "audio/mpeg",
    "audio/wav",
    "audio/mp4",
    "audio/x-m4a",
    "video/mp4",
    "video/quicktime",
    "video/webm",
}

MAX_UPLOAD_BYTES = 500 * 1024 * 1024

_MIME_BY_EXT = {
    ".mp3": "audio/mpeg",
    ".wav": "audio/wav",
    ".m4a": "audio/mp4",
    ".mp4": "video/mp4",
    ".mov": "video/quicktime",
    ".webm": "video/webm",
}


def sanitize_file_name(file_name: str) -> str:
    base = Path(file_name).name or "upload"
    return "".join(c if c.isalnum() or c in "._-" else "_" for c in base)


def source_storage_path(user_id: str, project_id: str, file_name: str) -> str:
    safe = sanitize_file_name(file_name)
    return f"{user_id}/{project_id}/source/{safe}"


def extracted_audio_storage_path(user_id: str, project_id: str) -> str:
    return f"{user_id}/{project_id}/extracted/audio.wav"


def source_file_type(media_type: str) -> str:
    return "source_audio" if media_type == "audio" else "source_video"


def infer_mime_type(file_name: str) -> str | None:
    ext = Path(file_name).suffix.lower()
    return _MIME_BY_EXT.get(ext)
